### PR TITLE
Preventing users from buying already gifted subscriber items

### LINKED ISSUE #7712: Unclaimed Mystery Items are purchasable with hourglasses
If a subscriber leaves the gift in the mystery box, there is nothing preventing them from purchasing the item in the time traveler shop.

### PR DESCRIPTION
**Pull request by Hus274**
closes #7712 
### Changes

Adding a check of the users unopened subscriber gifts when finding the subscriber items that the user doesn't have.

---

UUID: 2a4d30ee-4833-420e-be7c-ad6a6e261c5f

### CODE REVIEW

File: common/script/content/index.js [hunk 1]
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
   });
 });
 
-api.timeTravelerStore = function(owned) {
+api.timeTravelerStore = function(owned, unopenedGifts) {
+  var mystery = api.mystery;
   var ownedKeys;
   ownedKeys = _.keys((typeof owned.toObject === "function" ? owned.toObject() : void 0) || owned);
+  ownedKeys = _.union( ownedKeys, unopenedGifts);
```
Comment (crookedneighbor) [inline common/script/content/index.js]: Extra space

**Comment by crookedneighbor** [comment]
This behavior needs test coverage.

File: common/script/content/index.js [hunk 2]
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
   });
 });
 
-api.timeTravelerStore = function(owned) {
+api.timeTravelerStore = function(owned, unopenedGifts) {
+  var mystery = api.mystery;
```
Comment (khipkin) [inline common/script/content/index.js]: this value is not used

**Comment by crookedneighbor** [comment]
Hmmm... there's some general refactoring I want to do on these functions, so I'm going to take this over. Good work on getting it started. Looks like it solves the issue.

**Comment by Hus274** [comment]
Ok sounds good!

File: common/script/content/index.js [hunk 3]
```diff
@@ -38,9 +38,11 @@ _.each(api.mystery, function(v, k) {
   });
 });
 
-api.timeTravelerStore = function(owned) {
+api.timeTravelerStore = function(owned, unopenedGifts) {
```
Comment (crookedneighbor) [inline common/script/content/index.js]: Would it be a simpler api to just pass in the user here?

**Comment by crookedneighbor** [comment]
There are still a few pending comments on here, but it looks mostly 👍  to me.

**Comment by Hus274** [comment]
You mentioned above that you were going to finish this up, would you like me to do it instead?

**Comment by crookedneighbor** [comment]
Whoops. Totally forgot. Go ahead and finish it. I'm clearly not doing it :smile:

**Comment by Hus274** [comment]
Haha no worries, I can finish it up!

**Comment by crookedneighbor** [comment]
@Hus274 Status on this?

**Comment by Hus274** [comment]
@crookedneighbor sorry I've been busy the last week or so.  Should have some time to finish it up this this week. Will plan for either tomorrow or Tuesday.

**Comment by crookedneighbor** [comment]
No worries. Just wanted to make sure you weren't waiting on me for anything.

File: common/script/content/index.js (hunk 3 above)
Comment (Hus274) [reply common/script/content/index.js]: yeah that makes sense

File: common/script/content/index.js (hunk 2 above)
Comment (Hus274) [reply common/script/content/index.js]: fixed

File: common/script/content/index.js (hunk 1 above)
Comment (Hus274) [reply common/script/content/index.js]: fixed

**Review by Hus274 — COMMENTED** [review]
Just going to add up tests here and then I'll resubmit this.

**Comment by Hus274** [comment]
I've committed the changes that have been requested. I'm not sure where to put a test for the timeTravelerStore method as it's in index.js. Any thoughts?

**Comment by TheHollidayInn** [comment]
@Hus274 Is it possible to place this function somewhere else? If it must stay, you can put the test here: https://github.com/HabitRPG/habitica/tree/develop/test/content.

File: website/views/options/inventory/time-travelers.jade [hunk 4]
```diff
@@ -21,7 +21,7 @@
     .col-md-12
       li.customize-menu.inventory-gear
         menu.pets-menu(label='{{::category.text}}', ng-repeat='category in timeTravelersCategories')
-          div(ng-repeat='item in category.items', ng-if='category.identifier === "pets" || category.identifier === "mounts" ? !user.items[category.identifier][item.key] : !user.items.gear.owned[item.key]')
+          div(ng-repeat='item in category.items', ng-if='category.identifier === "pets" || category.identifier === "mounts" ? !user.items[category.identifier][item.key] : (!user.items.gear.owned[item.key] && user.purchased.plan.mysteryItems.indexOf(item.key) == -1)')
```
Comment (TheHollidayInn) [inline website/views/options/inventory/time-travelers.jade]: We should move this if logic to the controller.

**Comment by Hus274** [comment]
@TheHollidayInn I moved the complicated if logic to the controller and removed the ternary operator for clarity. Definitely looks better. I also moved the time traveler's logic into it's own file and added tests around the behavior. Let me know if there's anything else!

**Review by TheHollidayInn — APPROVED** [review]

**Pull request merged**